### ISSUE TITLE
undefined extends all types when accessing a generic in a mapped typed

### ISSUE THREAD

**Issue by squirly** [Microsoft/TypeScript#27470]
**TypeScript Version**: `3.2.0-dev.20180929` (First noticed in `3.1.1`)

**`strictNullChecks` must be on** [otherwise this is expected behaviour](https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#intersecting-with-nullundefined-reduces-to-nullundefined-outside-of-strictnullchecks).

It seems that an issue has been introduced when accessing attributes of a mapped type in a conditional type.

**Search Terms:**
conditional type undefined extends
mapped type undefined extends
conditional type in mapped type access

**Code**

```ts
type UndefinedKeys<T extends Record<string, any>> = {
  [K in keyof T]: undefined extends T[K] ? K : never
};

type MyType = {a: string, b: string | undefined}

type Result1 = UndefinedKeys<MyType>;

const a1: Result1['a'] = 'a'; // $ExpectError
      ~~  There should be an error here
const b1: Result1['b'] = 'b';

// manually mapping the type, error is show as expected
type Result2 = {
  a: undefined extends MyType['a'] ? 'a' : never;
  b: undefined extends MyType['b'] ? 'b' : never;
};

const a2: Result2['a'] = 'a'; // $ExpectError
const b2: Result2['b'] = 'b';
```

To simplify this even more line 5 can be changed to
```ts
type MyType = { a: string, b: undefined };
```

**Expected behavior:**

1. The type `Result1` should be `{a: never; b: 'b'}`.
1. Line 9 should have an error: `Type '"a"' is not assignable to type 'never'.`

**Actual behavior:**

1. The type `Result1` is `{a: 'a'; b: 'b'}`.
1. Line 9 has no error.

**Playground Link:** 

| **TypeScript Version** | Playground Link |
|---|---|
| 2.8.1 | [Playground](https://agentcooper.github.io/typescript-play/?ts=2.8.1#code/C4TwDgpgBAqgdgEwgMwJZwgg0hEBnAHgBUoIAPYCRPKAJQgGMB7AJwQL2BfQHMAaKAEM4IAHyioAXigBvAFBQoAbSxR0UANa4myKEQC6ALigBXRCnSZSFKghpEV+qAH4oq4xgBuEFnIC+ANxycqCQUACyIETg0NIygsac3HD8UABGiVy8UAA+puZoGAh+waHQ9HgmADbAAIxSsAWW2LiEkdGQokFyzHCcQrXGFdV1SgDkgmNO0hNjAVAA9AtQACQAomSQDMBrLCysPUx9wOmDdBCVNbXjaVMNY7fdS1AAtsImglVVIK+CYGDZYAAC2gZRCMXOl2AACYGvJFAl8khClZyJRqBEojFxpMnK5ZlAPBBvCwgooMkiLEVrOi7JiOhAbnd8bdCVAvD4goFgr1+oJoUMLiNoTi7jNJvNnutNowdnsDryTmkBZDhUzplAHnM5EA) |
| 3.0.3 | [Playground](https://agentcooper.github.io/typescript-play/#code/C4TwDgpgBAqgdgEwgMwJZwgg0hEBnAHgBUoIAPYCRPKAJQgGMB7AJwQL2BfQHMAaKAEM4IAHyioAXigBvAFBQoAbSxR0UANa4myKEQC6ALigBXRCnSZSFKghpEV+qAH4oq4xgBuEFnIC+ANxycqCQUACyIETg0NIygsac3HD8UABGiVy8UAA+puZoGAh+waHQ9HgmADbAAIxSsAWW2LiEkdGQokFyzHCcQrXGFdV1SgDkgmNO0hNjAVAA9AtQACQAomSQDMBrLCysPUx9wOmDdBCVNbXjaVMNY7fdS1AAtsImglVVIK+CYGDZYAAC2gZRCMXOl2AACYGvJFAl8khClZyJRqBEojFxpMnK5ZlAPBBvCwgooMkiLEVrOi7JiOhAbnd8bdCVAvD4goFgr1+oJoUMLiNoTi7jNJvNnutNowdnsDryTmkBZDhUzplAHnM5EA) |
| 3.1.1 | [Playground](http://www.typescriptlang.org/play/#src=type%20UndefinedKeys%3CT%20extends%20Record%3Cstring%2C%20any%3E%3E%20%3D%20%7B%0D%0A%20%20%5BK%20in%20keyof%20T%5D%3A%20undefined%20extends%20T%5BK%5D%20%3F%20K%20%3A%20never%0D%0A%7D%3B%0D%0A%0D%0Atype%20MyType%20%3D%20%7Ba%3A%20string%2C%20b%3A%20string%20%7C%20undefined%7D%0D%0A%0D%0Atype%20Result1%20%3D%20UndefinedKeys%3CMyType%3E%3B%0D%0A%0D%0Aconst%20a1%3A%20Result1%5B'a'%5D%20%3D%20'a'%3B%20%2F%2F%20%24ExpectError%0D%0Aconst%20b1%3A%20Result1%5B'b'%5D%20%3D%20'b'%3B%0D%0A%0D%0A%2F%2F%20manually%20mapping%20the%20type%0D%0Atype%20Result2%20%3D%20%7B%0D%0A%20%20a%3A%20undefined%20extends%20MyType%5B'a'%5D%20%3F%20'a'%20%3A%20never%3B%0D%0A%20%20b%3A%20undefined%20extends%20MyType%5B'b'%5D%20%3F%20'b'%20%3A%20never%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20a2%3A%20Result2%5B'a'%5D%20%3D%20'a'%3B%20%2F%2F%20%24ExpectError%0D%0Aconst%20b2%3A%20Result2%5B'b'%5D%20%3D%20'b'%3B%0D%0A) <---- ENABLE `strictNullChecks` IN OPTIONS |
| 3.2.0-dev.20180929 | Same error as 3.1.1 |

**Related Issues:**
Potentially: https://github.com/Microsoft/TypeScript/issues/26942

Initially thought it was an issue with [Distributive conditional types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types) but the Generic type is in the assignable to position.

**Comment by jack-williams** [comment]
The issue seems to be the constraint on `T`. The following has the intended behavior:
```ts
type UndefinedKeys<T> = {
  [K in keyof T]: undefined extends T[K] ? K : never
};
```

The index access expressions `T[K]` seems to be pulling in the constraint where all property types extend `any`,  and therefore `undefined` is related.

I think this is a minimal repro (using arrays).
```ts
type X<T extends any> = undefined extends T ? true : false;
type Y<T extends any[]> = undefined extends T[number] ? true : false;

type A = X<number>; // ignores the `any` constraint and returns false
type B = Y<number[]>; // returns true;
```

**Comment by ahejlsberg** [comment]
This one is rather involved. We currently have the following rule:

* A type `S` is related to a type `T[K]` if `S` is related to `C`, where `C` is the base constraint of `T[K]`.

This rule is unsound because `T[K]` may be a subtype of `C` upon instantiation, but we nonetheless need it for several of reasons. One key reason is the fact that we don't use higher order types for property accesses through type variables. For example, for a property access `this.xxx` in a method of a class, we resolve to the declared type of `xxx`, not type `this["xxx"]`, even though the latter is more precise. Effectively, we resolve to the _constraint_ of `this["xxx"]`. We do this for reasons of backwards compatibility and also to avoid drowning in generic types. But it would be odd indeed if the expression `this.xxx` was not assignable to the type `this["xxx"]`--hence the rule that allows the assignment if it matches the constraint.

Now, in 3.1 we fixed several issues relating to us not examining all constraints of a type. Because of those fixes we now drill all the way down and find the `any` constraint for `T[K]` in your example above. And since anything is assignable to `any` we effectively turn off type checking. To wit, none of the following assignments generate errors:

```ts
function test<T extends Record<string, any>, K extends keyof T>(t: T, k: K) {
    t[k] = 42;
    t[k] = "hello";
    t[k] = undefined;
}
```

That's not good. But the question is how to fix this without breaking the `this.xxx` scenario I mentioned above (and other similar ones).

So far I think the best contender is to modify the rule above as follows:

* A type `S` is related to a type `T[K]`, _where `T` and `K` aren't both type variables_, if `S` is related to `C`, where `C` is the base constraint of `T[K]`.

In other words, we should limit the unsoundness to situations where one or the other part of the indexed access is an actual zero-order type, but forbid it when both are higher order (since at that point it becomes _too_ unsound).

I ran the experiment on our test suites and there just a few changes, all of which are reasonable. Even better, the RWC suites show no baseline changes at all. I will put this up in a PR.

**Comment by jack-williams** [comment]
> And since anything is assignable to any we effectively turn off type checking. To wit, none of the following assignments generate errors

FWIW I don't think it is because type-checking is turned off, but because `any` is a top type (I don't know if you consider this a distinction). You get the same behavior with `unknown`. 
```ts
type UndefinedKeys<T extends Record<string, unknown>> = {
  [K in keyof T]: undefined extends T[K] ? K : never
};
type MyType = {a: string, b: string | undefined}
type Result1 = UndefinedKeys<MyType>;
const a1: Result1['a'] = 'a'; // no error
```
The conditional type is essentially a type-level version of this:
```ts

const x: {a: string} = { a: "a" };
const baseConstraint: Record<string, unknown> = x;
baseConstraint.a = undefined; // unsound
```
> A type S is related to a type T[K], where T and K aren't both type variables, if S is related to C, where C is the base constraint of T[K].

Am I right in say that this would not address: 

`type Y<T extends any[]> = undefined extends T[number] ? true : false;`? 

(This isn't a complaint, just wanting to know if I read the rule ok).

> That's not good. But the question is how to fix this without breaking the this.xxx scenario I mentioned above (and other similar ones).

From an uniformed/outside perspective it seems like the problem trying to be solved in the initial post is to constrain `T` to some shape (a record, but it could be an array). The programmer only really cares about the outermost data constructor and the any is just there because it causes little friction. Though when you pull down the constraints the `any` starts causing problems and you can't control the variance.

I wonder whether it would be possible to have a special way of marking shapes for types where you really don't care about, or particularly, want their constituent constraints. Something like:
```ts
type UndefinedKeys<T extends Record<string, *>> = {
  [K in keyof T]: undefined extends T[K] ? K : never
};
```
`T extends Record<string, *>` can be read as: `T` should be some record but other than that we don't care about the type of the value, please don't pull the constraint down because I'm going to use `T[K]` in a contravariant way! `*` is basically an alias for `any` that doesn't get pulled out from constraints, or an inference where we discard the result.

**Comment by ahejlsberg** [comment]
> FWIW I don't think it is because type-checking is turned off, but because any is a top type (I don't know if you consider this a distinction). You get the same behavior with unknown.

That is indeed what I meant.

> I wonder whether it would be possible to have a special way of marking shapes for types where you really don't care about, or particularly, want their constituent constraints.

I would be more inclined to introduce a new type relationship (e.g. "soundly assignable") that excludes the unsound rule and then use that relationship when checking the `extends` clause of a conditional type. But we really try hard not to have too many type relationships because we end up consuming more time and memory resources.

**Comment by jack-williams** [comment]
@aleclarson Yes this seems to be related. The fix made only works for index accesses of the form `T[K]`, where both `T` and `K` are parameters; your code uses `T['a']`.

> Using "brackets on a mapped type" inside an "extends" clause lets `false extends true` evaluate to true.

This isn't 100% true in the sense that what is going on is that the condition `false extends T['a']` is being eagerly reduced using the constraint `Foo`. So the condition is being evaluated as `false extends boolean`, which correctly selects the true branch. The conditional type is being resolved before the function ever gets called.
